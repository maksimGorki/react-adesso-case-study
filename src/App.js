import "./css/style.css";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import React, { useEffect } from "react";
import { useDispatch } from "react-redux";

import Navbar from "./components/Navbar";
import HomePage from "./components/HomePage";
import UnitsPage from "./components/UnitsPage";
import UnitDetailPage from "./components/UnitDetailPage";

const actionTypes = {
  fetchData: "FETCH_DATA",
  filterData: "FILTER_DATA",
};

function App() {
  const dispatch = useDispatch();

  const url = "data/age-of-empires-units.json";
  useEffect(() => {
    dispatch({ type: actionTypes.fetchData, url: url });
  }, [url, dispatch]);

  return (
    <Router>
      <div className="wrapper">
        <Navbar />
        <Switch>
          <Route exact strict path="/" component={HomePage} />
          <Route exact strict path="/unitsPage" component={UnitsPage} />
          <Route
            exact
            strict
            path="/unitDetailPage/:unitname"
            component={UnitDetailPage}
          />
        </Switch>
      </div>
    </Router>
  );
}

export default App;
