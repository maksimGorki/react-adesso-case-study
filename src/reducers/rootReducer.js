const initialState = {
  fetchedUnits: [],
  modifiedUnitsByAge: [],
  modifiedUnitsByGoods: [],
  modifiedUnitsOverall: [],
};

const actionTypesAsync = {
  fetchDataAsync: "FETCH_DATA_ASYNC",
  filterDataAsync: "FILTER_DATA_ASYNC",
};

const ageTypes = {
  allAges: "All",
  darkAge: "Dark",
  feudalAge: "Feudal",
  castleAge: "Castle",
  imperialAge: "Imperial",
};

const rootReducer = (state = initialState, action) => {
  const newState = { ...state };

  switch (action.type) {
    case actionTypesAsync.fetchDataAsync:
      fetch(action.url)
        .then((response) => response.json())
        .then((data) => {
          newState.fetchedUnits = data.units;
          newState.modifiedUnitsByAge = data.units;
        });
      break;

    case actionTypesAsync.filterDataAsync:
      if (action.age !== undefined) {
        if (action.age === ageTypes.allAges) {
          newState.modifiedUnitsByAge = newState.fetchedUnits;
        } else {
          newState.modifiedUnitsByAge = newState.fetchedUnits.filter(
            (item) => item.age === action.age
          );
        }
      }

      newState.modifiedUnitsByGoods = newState.modifiedUnitsByAge;
      if (action?.wood?.checked) {
        newState.modifiedUnitsByGoods = newState.modifiedUnitsByGoods.filter(
          (item) =>
            item?.cost?.Wood <= action.wood.end &&
            item?.cost?.Wood >= action.wood.start
        );
      }
      if (action?.food?.checked) {
        newState.modifiedUnitsByGoods = newState.modifiedUnitsByGoods.filter(
          (item) =>
            item?.cost?.Food <= action.food.end &&
            item?.cost?.Food >= action.food.start
        );
      }
      if (action?.gold?.checked) {
        newState.modifiedUnitsByGoods = newState.modifiedUnitsByGoods.filter(
          (item) =>
            item?.cost?.Gold <= action.gold.end &&
            item?.cost?.Gold >= action.gold.start
        );
      }
      break;

    default:
      break;
  }
  newState.modifiedUnitsOverall = newState.modifiedUnitsByGoods;
  return newState;
};

export default rootReducer;
