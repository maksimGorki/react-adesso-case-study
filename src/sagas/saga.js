import { takeLatest, put, all, delay } from "redux-saga/effects";

const actionTypes = {
  fetchData: "FETCH_DATA",
  filterData: "FILTER_DATA",
};

const actionTypesAsync = {
  fetchDataAsync: "FETCH_DATA_ASYNC",
  filterDataAsync: "FILTER_DATA_ASYNC",
};

function* fetchDataAsync(action) {
  yield put({ type: actionTypesAsync.fetchDataAsync, url: action.url });
}

function* filterDataAsync(action) {
  yield delay(100); // debouncing filter
  yield put({
    type: actionTypesAsync.filterDataAsync,
    age: action.currentAge,
    wood: action.wood,
    food: action.food,
    gold: action.gold,
  });
}

export function* rootSaga() {
  yield all([
    takeLatest(actionTypes.filterData, filterDataAsync),
    takeLatest(actionTypes.fetchData, fetchDataAsync),
  ]);
}

export default rootSaga;
