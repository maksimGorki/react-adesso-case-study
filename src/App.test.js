import React from "react";
import ReactDOM from "react-dom";
import App from "./App";

import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

import rootReducer from "./reducers/rootReducer";
import { Provider } from "react-redux";
import { createStore, applyMiddleware } from "redux";
import createSagaMiddleware from "redux-saga";
import { rootSaga } from "./sagas/saga";

const sagaMiddleware = createSagaMiddleware();
const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));
sagaMiddleware.run(rootSaga);

afterEach(cleanup);

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(
    <Provider store={store}>
      <App />
    </Provider>,
    div
  );
});

it("matches snapshot", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <App />
      </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
