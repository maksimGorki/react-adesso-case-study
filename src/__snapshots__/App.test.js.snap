// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="wrapper"
>
  <div
    className="navbar"
  >
    <a
      aria-current="page"
      className="navbarItem active"
      href="/"
      onClick={[Function]}
      style={Object {}}
    >
      Home
    </a>
    <a
      aria-current={null}
      className="navbarItem"
      href="/unitsPage"
      onClick={[Function]}
    >
      Units
    </a>
  </div>
  <div
    className="homePageContainer"
  >
    <div
      className="homePageTitleName"
    >
      Home Page
    </div>
    <div
      className="homePageImageBox"
    >
      <img
        alt="Age of Empires Img"
        src="age_of_empires2_the_age_of_kings.jpg"
      />
    </div>
  </div>
</div>
`;
